stempy/image.cpp

```diff
@@ -295,36 +295,59 @@ std::vector<double> getContainer(const STEMImage& inImage, const int numBins)
   return container;
 }
 
-// function that computes histogram for all the STEM images
+// function that computes numHist histograms for input STEM image
 // each histogram is a vector<int>
-std::vector<int> createSTEMHistogram(const STEMImage& inImage,
-                                     const int numBins,
-                                     const std::vector<double> bins)
+std::vector<std::vector<int>> createSTEMHistogram(
+  const STEMImage& inImage, const int numHist, const int numBins,
+  const std::vector<double>& bins)
 {
-  // initialize output
-  std::vector<int> frequencies(numBins, 0);
-
   // STEMImage info
   int width = inImage.width;
   int height = inImage.height;
   auto curData = inImage.data;
 
-  // get a histrogram
+  // initialize output - the input image can have numHist histograms
+  // where each corresponds to a piece of original image
+  std::vector<std::vector<int>> allFrequencies;
+
+  // initialize total data with potential zero-padding
+  int numData = width * height;
+  if (numData % numHist != 0) {
+    numData += numData % numHist;
+  }
+  vector<int> totalData(numData, 0);
+  // zero-padded area (numData > width*height) remains 0
   for (int i = 0; i < width * height; i++) {
-    auto value = curData[i];
-    // check which bin it belongs to
-    for (int j = 0; j < numBins; j++) {
-      if (value >= bins[j] && value < bins[j + 1]) {
-        ++frequencies[j];
+    totalData[i] = curData[i];
+  }
+
+  // length of individual data
+  int subLength = numData / numHist;
```

Comment:
I was taking a quick look at what happens here if `(width * height) % numHist != 0`. I printed out some debug output (11 histograms for this example):
```
width * height is: 25600
numData is: 25603
numHist is: 11
subLength is: 2327
```

Correct me if I'm wrong, but it looks like this will not cover all of the data? `subLength * numHist` is `25597`, which is 3 less than `width * height`. So it looks to me that in the loops below, the last three values of the image will not be placed in a bin?

Reply:
Yeah it is indeed a bug. The line of code that updates `numData` should actually be
`numData += (numHist - numData % numHist);`
instead of
`numData += numData % numHist;`

Thank you very much for testing it out!

Reply:
That is good! But we do also want to make sure that all of the data points get placed in bins, right? If `subLength * numHist` is 25597, doesn't that mean that the last 3 values (`width * height == 25600`) do not get placed in any bins? Perhaps you should add something to make sure that all of the leftover values (which will be equal to `(width * height) % numHist`) get placed in bins as well.

Reply:
The way I solved the issue you mentioned was by appending zeros to the "left-over values". Continue with your `numHist=11` example, the last 3 values will be appended with 8 zeros and all together be used to generate the histogram. Adding zeros does not affect anything because for our data, the min value is always positive (actually super large, which can be double-checked from the lower bound of the generated histogram), so having some additional zeros would not change the correctness of the histogram.

Reply:
Oh yes, I see that now. Sounds good!